x-pack/legacy/plugins/lens/public/types.ts

```diff
@@ -59,6 +63,8 @@ export interface Datasource<T = unknown, P = unknown> {
 
   toExpression: (state: T) => Ast | string | null;
 
+  getMetaData: (state: T) => DatasourceMetaData;
```

Comment:
I traced through the entire change that this introduces and I don't understand why it's here.

1. The saved object gets a list of strings indicating which index patterns are used in the visualization
2. The embeddable is using this to load instances of the Index Pattern Service on initialize
3. The embeddable doesn't seem to use the loaded index pattern service at all

I'm not sure what this code is doing, it seems redundant. Am I missing something?

Reply:
This meta data is used by the dashboard app to configure the filter bar. It's basically about this combo box:
<img width="476" alt="Screenshot 2019-07-22 at 09 25 31" src="https://user-images.githubusercontent.com/1508364/61613867-b0883600-ac62-11e9-9f49-8ce421b48134.png">

I agree the control flow is not super straight forward here. It is used here: https://github.com/elastic/kibana/pull/41361/files#diff-2da2f102b212621f4267b4cc25bd5810R69 . The embeddable has an abstract base class which passes the used index patterns as "output" to the embeddable container (in this case the dashboard container) which in turn passes them to the filter bar.

Reply:
Okay, finally followed all the logic of this through. You may not be aware but the `SearchBar` component is able to take a string as input and generate an instance of IndexPatternService from that, so we probably don't need to return anything but the ID of the index pattern here.

Reply:
Thought about this some more and followed the dashboard-related code more. I'm fine with loading the index pattern service here, it is the right thing to do when dealing with dashboards and the search bar